### ISSUE TITLE
Reclaiming more storage after delete+compact+vacuum

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#4934]
## tl;dr

The Zed lake maintains history about stored data on the assumption users may want to [time travel](https://zed.brimdata.io/docs/tutorials/zed#time-travel) to revisit the data as it looked in a prior state. The amount of storage consumed to maintain that history may become significant in some circumstances such as when there's been lots of commits (small data loads, granular deletes, etc.) When users decide they only care about data in its most recent state, command like `zed manage` and `zed vacuum` can reclaim lots of storage by compacting/deleting underlying data objects that will no longer be referenced. However, some commit/snapshot history is still currently left behind. There should be commands that can help users safely reclaim that storage as well.

## Details

At the time this issue is being opened, Zed it at commit afba255.

A community zync user recently asked the following question:

> HI,  in some ZED server, we have some pool with high volume of data on disk but only a few row (under 5) in it.  We pass the vacuum command and manage command to clean/compact it but still consume space on disk. 
> All the data consumed on disk are in the commits folder, Inside the folder, it have a lots of  .zng and .snap.zng files.
> How can I clean data on disk unused but the pool ?

Here's an in-house repro that shows the effects similar to what's being described here. To simulate the extreme example of single-row commits, we've got a "record poster" tool that generates synthetic data one JSON object of a time. A sample of a ew such commits:

```
Posting: {"ts": "2023-12-13T19:04:21.437554Z", "message": "jocoseriosity enactment forevalue enemylike Flustra tabular"}
Posting: {"ts": "2023-12-13T19:04:21.441720Z", "message": "dereliction donnism internarial"}
Posting: {"ts": "2023-12-13T19:04:21.446592Z", "message": "serenata salpingorrhaphy checkrope toothlessly oligochaete"}
...
```

A `zed serve` is started on an empty lake directory and the tool is used to post 10k records. At the end of that, our lake directory is ~156 MB in size. Broken down by what's under the directories that make up the specific pool:

```
$ du -sh *
 39M	branches
 39M	commits
 78M	data

$ for dir in *; do   echo -n "$dir:"; find $dir | wc -l; done
branches:   10005
commits:   10001
data:   20001
```

To reproduce what the user is describing, I delete all but the five most recently-posted records, then compact and vacuum in an attempt to drop stored remnants of what came before. At the end that leaves me with just the 5 records I care about.

```
$ zed -version
Version: v1.11.1-26-gafba2551

$ zed delete -where "ts < $(zed query -z 'from TestPool | head 5 | tail 1 | yield ts')"
2ZV8Gddh1gdegCTJJBv5G9RFcbi delete committed

$ zed manage
{"level":"info","ts":1702495810.833336,"logger":"pool","msg":"updating pool","name":"TestPool","id":"2ZV83duaJl7mdArqRNiG9mB2XFp","branch":"main","interval":60,"vectors":false}
{"level":"info","ts":1702495810.871395,"logger":"pool","msg":"compaction completed","name":"TestPool","id":"2ZV83duaJl7mdArqRNiG9mB2XFp","branch":"main","interval":60,"vectors":false,"runs_found":1,"objects_compacted":5,"vectors_created":0}

$ zed vacuum -f
vacuumed 10000 objects

$ zed query 'from TestPool | count()'
5(uint64)
```

We can now see that nearly all the storage has been reclaimed from `data/`, but plenty is still consumed by `branches/` and `commits/`.

```
$ du -sh *
 39M	branches
 40M	commits
8.0K	data

$ for dir in *; do   echo -n "$dir:"; find $dir | wc -l; done
branches:   10007
commits:   10006
data:       3
```

In terms of what the user sees if they go looking for what's consuming it:

```
$ ls -l branches/
total 80048
-rw-r--r--  1 phil  staff  118 Dec 13 11:28 1.zng
-rw-r--r--  1 phil  staff  121 Dec 13 11:28 10.zng
-rw-r--r--  1 phil  staff  121 Dec 13 11:28 100.zng
-rw-r--r--  1 phil  staff  121 Dec 13 11:28 1000.zng
-rw-r--r--  1 phil  staff  121 Dec 13 11:28 10000.zng
-rw-r--r--  1 phil  staff  121 Dec 13 11:28 10001.zng
-rw-r--r--  1 phil  staff  121 Dec 13 11:29 10002.zng
...

$ ls -l commits/
...
-rw-r--r--  1 phil  staff     384 Dec 13 11:28 2ZV8ABb5lXFR1IDFyZKK7fDf2F8.zng
-rw-r--r--  1 phil  staff     384 Dec 13 11:28 2ZV8ABcqNJI3YmscNEPcLVx8EzE.zng
-rw-r--r--  1 phil  staff     385 Dec 13 11:28 2ZV8ABdgmYgpfVejoOrMxpYvnaH.zng
-rw-r--r--  1 phil  staff     384 Dec 13 11:28 2ZV8ABdu1Sbzkkmx4d0f0aiquGp.zng
-rw-r--r--  1 phil  staff     386 Dec 13 11:28 2ZV8ABfQFUCI1njUDxpWK1QRBtP.zng
-rw-r--r--  1 phil  staff     349 Dec 13 11:30 2ZV8Gddh1gdegCTJJBv5G9RFcbi.snap.zng
-rw-r--r--  1 phil  staff  507773 Dec 13 11:29 2ZV8Gddh1gdegCTJJBv5G9RFcbi.zng
-rw-r--r--  1 phil  staff     178 Dec 13 11:30 2ZV8IzgXEYvHal5sI28bGN9eVGH.snap.zng
-rw-r--r--  1 phil  staff     696 Dec 13 11:30 2ZV8IzgXEYvHal5sI28bGN9eVGH.zng
```

In the specific case of the community zync user, they actually had more snapshots (`*.snap.zng`) than shown here.

![image](https://github.com/brimdata/zed/assets/5934157/18bb90dc-c3b7-4594-aab3-459cea71483e)

The [Zed Lake Format doc](https://zed.brimdata.io/docs/lake/format) goes into some more detail of the role of these different files. The bottom line is that if the user in this example truly only cares about the 5 most recent records and is ok losing the other history, there's not currently a `zed` command that does that automatically. And when that's implemented, it's TBD whether that would be done as a side effect of the `vacuum` command or a separate command.

In the meantime, after getting a refresher from the Dev team about the role of these different files, I was reminded of one short-term measure that users could safely take on their own to recoup some of this storage. Specifically, the snapshot files `commits/*.snap.zng` can be deleted by hand. This is because each snapshot just represents a point-in-time summary of history that would otherwise have to be constructed by walking the many non-snapshot files in that `commits/` directory. Therefore if there's no need to revisit that history at peak performance, these files are no longer needed.

**Comment by philrz** [comment]
We had a group discussion about this one. More specifics TBD, but some notes of thoughts thus far:
1. @mattnibs thought this kind of file cleanup should probably be handled by `zed manage`
1. There was consensus that the snapshot files could be moved to their own directory
   1. Since finding snapshot files currently requires directory scanning, this should simplify that
   1. It would also make it easier to communicate the short-term workaround to users (i.e., "it's ok to delete everything under `snapshots/`)
1. Think more about what's in `branches/`
